### ISSUE TITLE
Review logic related to login/out and dis/connect

### ISSUE THREAD

**Issue by michaelsbradleyjr** [status-im/nim-status#300]
In PR #271, I left [this comment](https://github.com/status-im/nim-status/blob/feat/waku-integration-in-library/status/api/common.nim#L134) above `proc close*(self: StatusObject)`:
```
# this and logic around login/out and dis/connect needs to be reconsidered
```

At the time I wrote that, I had a hunch that there is code within both `api/` and `private/` that doesn't fully take into account state transitions that can occur re: login/out and dis/connect, especially re: asynchronous operations.

Yesterday, when experimenting with the example client re: the changes in #271, I found an example of that, albeit the code was some I wrote myself for that PR.

How to trigger it:
1. Change the code in `api/waku.nim` so that there is e.g. a 100 millisecond `sleepAsync` after processing each message in the history query response, and then recompile the example client. It can de done [like this](https://gist.github.com/michaelsbradleyjr/1d970c159595a6a2b7f2af3c33feab40).
2. `/login ...`.
3. Allow the client to connect and start displaying the messages retrieved via the automatic post-connect history query.
4. `/logout`

Notice that the messages in the history query response continue to be processed.

Now, that's not a problem or wrong behavior *per se*: the symmetric key for public chat messages is derived from information that's independent of the user db, i.e. `shortName` metadata tracked in instances of `ContentTopic` at the time a chat is joined.

However, imagine if the messages being processed were e.g. 1-1 or private group messages. After `/logout`, and with the user db closed, it's at most a wrong behavior and probably not possible to continue processing messages in a history query response.

In summary, we'll need to start taking a close look at our existing public APIs and ones added in the future, keeping in mind state transitions that may occur asynchronously in relation to async APIs. Where appropriate, we'll want to have explicit state checks that translate to error results or error signals.
